docs/design/pitr-mysql.md

```diff
@@ -0,0 +1,218 @@
+
+
+# Critical User Journeys
+
+This section describes the common critical user journeys about why and how the users may use the Point-in-time Recovery feature to recover their data.
+
+## OMG Moments
+
+Occasionally, catastrophic events can happen to the databases and result in service downtime. Here are some cases where the users would like to recover the database to a specific point in time.
+
+- **Drop Database**: the database is dropped accidentally by mistake.
+- **Schema Migration Failure**: A column is dropped, or its type is updated incorrectly. The table needs to be rolled back to the state before the migration. In the long term, we could polish the feature to provide the ability to revert the migration with minimal cost. Still, currently, we could leverage PITR to do the lift, with the limitation of losing incoming data after the migration.
+- **Buggy Application**: Bugs happen, and they may corrupt your data. The chances are that the user rolls out a new release of their service, only to find a bug in the application code that deletes essential data. It is best to roll back the service and recover the database to the state just before the release. Here’s when PITR kicks in.
+
+## Database Recovery
+
+When OMG moments happen, the user would like to recover the database to the latest correct state as soon as possible. This is where the Point-in-time Recovery feature kicks in.
+
+As long as the instance fulfills all the [requirements](#requirements) and enables a backup policy in the Bytebase environment that contains this instance, the user is lucky enough to have a PITR option. The user could then choose those databases they want to recover, and select the point of time, then Bytebase will recover the databases to precisely the state of that point of time.
+
+What’s more convenient is that if the user used Bytebase to do the wrong schema migration, then Bytebase will record the exact time before the schema migration and provide a one-click Point-in-time Recovery experience.
+
+
+# Design
+
+## Point-in-time Recovery Process
+
+### Full Backup
+
+Full backup follows what `mysqldump` does with `--source-data` flag. `SHOW MASTER STATUS` gives the coordinate (MySQL binlog file name and position) of a backup. This provides us with the starting position to apply binary logs in the PITR recovery step. Because the `SHOW MASTER STATUS` statement is not transactional, we have to put [`FLUSH TABLES tbl_names... WITH READ LOCK`](https://dev.mysql.com/doc/refman/8.0/en/flush.html#flush-tables-with-read-lock-with-list) to block writes to figure out the precise binary log coordinate in the transaction of taking database backup. As a side-effect, writes will be blocked for a short period between steps 1 and 4. By the way, we have to make sure the tables are unlocked regardless of any failure.
+
+1. `FLUSH TABLES tbl_names... WITH READ LOCK`.
+2. `SHOW MASTER STATUS`. Bytebase will record the binary log coordinates and timestamp.
```

Comment:
Is there a chance the process stops somehow after 2, will this cause we to hold the lock forever?

Reply:
I feel we need to mention the timer. How we implement the timer can be an implementation detail.

Reply:
Added description here https://github.com/bytebase/bytebase/pull/1220/files#diff-ec65251dcc3acb23d2cbc09c75c07268d48b1236854fc9fca524d6a4f5fd6b7dR72